### PR TITLE
[Merged by Bors] - feat: generalise results about Lie module weight spans from Abelian to nilpotent Lie algebras

### PR DESCRIPTION
**Pull request by ocfnash**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Lie/Killing.lean [hunk 1]
```diff
@@ -438,8 +473,8 @@ section Field
 open LieModule FiniteDimensional
 open Submodule (span)
 
-lemma LieModule.traceForm_eq_sum_finrank_nsmul_mul
-    [IsLieAbelian L] [IsTriangularizable K L M] (x y : L) :
+lemma LieModule.traceForm_eq_sum_finrank_nsmul_mul [LieAlgebra.IsNilpotent K L]
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/Killing.lean:476]: The weakening of hypotheses here is the main point of these changes.

File: Mathlib/Algebra/Lie/Weights/Linear.lean [hunk 2]
```diff
@@ -101,16 +111,87 @@ lemma zero_lt_finrank_weightSpace {χ : L → R} (hχ : weightSpace M χ ≠ ⊥
...
+      ← trace_comp_toEndomorphism_weight_space_eq, LinearMap.comp_apply, LieHom.coe_toLinearMap,
+      LieHom.map_lie, Ring.lie_def, map_sub, LinearMap.trace_mul_comm, sub_self]
 
 end FiniteDimensional
 
+variable [LieAlgebra.IsNilpotent R L] [LinearWeights R L M] (χ : L → R)
+
+/-- A type synonym for the `χ`-weight space but with the action of `x : L` on `m : weightSpace M χ`,
+shifted to act as `⁅x, m⁆ - χ x • m`. -/
+def shiftedWeightSpace := weightSpace M χ
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/Weights/Linear.lean:135]: Naming suggestions (always welcome) especially welcome here!

File: Mathlib/Algebra/Lie/Killing.lean [hunk 3]
```diff
@@ -411,6 +441,11 @@ lemma restrictBilinear_killingForm (H : LieSubalgebra R L) :
     H.restrictBilinear (killingForm R L) = LieModule.traceForm R H L :=
   rfl
 
+@[simp] lemma ker_traceForm_of_isCartanSubalgebra_eq_bot
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Killing.lean:444]: Shouldn't the name be
```suggestion
@[simp] lemma ker_traceForm_eq_bot_of_isCartanSubalgebra
```
I guess the same applies to `ker_restrictBilinear_of_isCartanSubalgebra_eq_bot` which must have slipped through during the review of an earlier PR.

File: Mathlib/Algebra/Lie/Weights/Linear.lean [hunk 4]
```diff
@@ -101,16 +111,87 @@ lemma zero_lt_finrank_weightSpace {χ : L → R} (hχ : weightSpace M χ ≠ ⊥
...
+
+namespace shiftedWeightSpace
+
+private lemma aux [Nontrivial (shiftedWeightSpace R L M χ)] : weightSpace M χ ≠ ⊥ :=
+  (LieSubmodule.nontrivial_iff_ne_bot _ _ _).mp <| inferInstanceAs <| Nontrivial <|
+    shiftedWeightSpace R L M χ
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Weights/Linear.lean:141]: Does this work?
```suggestion
private lemma aux [h : Nontrivial (shiftedWeightSpace R L M χ)] : weightSpace M χ ≠ ⊥ :=
  (LieSubmodule.nontrivial_iff_ne_bot _ _ _).mp h
```

File: Mathlib/Algebra/Lie/Killing.lean (hunk 3 above)
Comment (ocfnash) [reply Mathlib/Algebra/Lie/Killing.lean]: Good point, done.

File: Mathlib/Algebra/Lie/Weights/Linear.lean (hunk 4 above)
Comment (ocfnash) [reply Mathlib/Algebra/Lie/Weights/Linear.lean]: Better, thanks!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**